pkg/promtail/targets/syslogparser/syslogparser.go

```diff
@@ -0,0 +1,71 @@
+package syslogparser
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+
+	"github.com/influxdata/go-syslog/v2"
+	"github.com/influxdata/go-syslog/v2/octetcounting"
+	"github.com/influxdata/go-syslog/v2/rfc5424"
+)
+
+// ParseStream parses a rfc5424 syslog stream from the given Reader, returning a channel
+// emitting the parsed messages. The parser automatically detects octet counting.
+// The channel is closed on EOF or unrecoverable errors.
+func ParseStream(r io.Reader) <-chan *syslog.Result {
```

Comment:
Instead of creating channel and goroutine, can we instead make use this signature:

```suggestion
func ParseStream(r io.Reader, callback func(res *syslog.Result)) error {
```

and pass all parsed messages directly to `callback` function? I think we don't need extra goroutine here, as there is already goroutine for this connection, which is currently just passing messages from one channel to another. It can do the parsing as well.

Reply:
I totally agree! Done :)